services/src/floodlight-modules/src/main/java/org/openkilda/floodlight/kafka/Consumer.java

```diff
@@ -81,4 +115,52 @@ protected void handle(ConsumerRecord<String, String> record) {
         logger.trace("received message: {} - {}", record.offset(), record.value());
         handlersPool.execute(handlerFactory.produce(record));
     }
+
+    @VisibleForTesting
+    static class KafkaOffsetRegistry {
+        private final KafkaConsumer<String, String> consumer;
+        private final long autoCommitInterval;
+        private final Map<TopicPartition, Long> partitionToUncommittedOffset = new HashMap<>();
+        private long lastCommitTime;
+
+        KafkaOffsetRegistry(KafkaConsumer<String, String> consumer, long autoCommitInterval) {
+            this.consumer = requireNonNull(consumer);
+            checkArgument(autoCommitInterval > 0, "autoCommitInterval must be positive");
+            this.autoCommitInterval = autoCommitInterval;
+
+            lastCommitTime = System.currentTimeMillis();
+        }
+
+        /**
+         * Add the record's offset to the registry and perform a commit
+         * if more than autoCommitInterval ms passed since the last commit.
+         */
+        void addAndCommit(ConsumerRecord<String, String> record) {
+            TopicPartition partition = new TopicPartition(record.topic(), record.partition());
+            partitionToUncommittedOffset.put(partition, record.offset());
```

Comment:
Is the offset always grow? Should be check/filter out rewind attempts?

Reply:
I'd added the check.